### ISSUE TITLE
False-positive need-check-nil after string/table normalization with and/or

### ISSUE THREAD

**Issue by lewis6991** [EmmyLuaLs/emmylua-analyzer-rust#1260]
## Description

Normalizing a `string|{src: string}` parameter with `and`/`or` makes the required `src` field appear nullable. The equivalent explicit `if` branch produces no diagnostics.

## Minimal reproduction

In an otherwise empty directory, create `.emmyrc.json` containing `{}` and `repro.lua`:

```lua
---@param spec string|{src: string}
local function normalize(spec)
  spec = type(spec) == 'string' and { src = spec } or spec
  return spec.src:upper()
end
return normalize
```

Run:

```sh
emmylua_check . --warnings-as-errors
```

Tested with `emmylua_check 0.25.1` on macOS arm64. No Neovim dependencies are needed.

## Actual behavior

Exit status 1, with these diagnostics on line 4:

```text
warning: expected `string` but found `string?`. [param-type-mismatch]
warning: spec.src may be nil [need-check-nil]
```

## Expected behavior

After normalization, `spec.src` is always a string:

- The string branch creates `{ src = spec }`, with `spec` narrowed to string.
- The table branch keeps the original table, whose `src` field is required.
- The newly constructed table is always truthy, so the `or` branch cannot replace it.

## Controls / workarounds

Replacing the assignment with this produces no diagnostics:

```lua
if type(spec) == 'string' then
  spec = { src = spec }
end
```

An explicit annotation of the complete expression also produces no diagnostics:

```lua
spec = (type(spec) == 'string' and { src = spec } or spec) --[[@as {src: string}]]
```

The inline and explicit-branch versions were also run with LuaJIT and returned identical results for string/table inputs, including empty strings.

Encountered in Neovim's `vim.pack` specification normalization. Possibly related to #921 (closed), but this reproduction involves reassigning the narrowed variable through an `and`/`or` expression.
